### PR TITLE
swift3 docker build speed boost base on image

### PR DESCRIPTION
**Pull request by csantanapr**
PG1 1347 🏃

### CODE REVIEW

**Comment by rabbah** [comment]
@paulcastro using the ibm image, reliable?

**Comment by rabbah** [comment]
Carlos, one bit to keep in mind is we want to run the proxy with python 3 soon so we might need to independently build and publish our own swift image. In any case having the action here built from a different base is the right idea in terms of getting the separation we want. 👍

**Comment by paulcastro** [comment]
Look good to me - 1404, etc matches us.  We had an issue with an older clang in the base image.  Upping the version was part of our slow-down.

**Comment by paulcastro** [comment]
One other comment, this container is Swift 3.0.1. Per issue #2079 I'd like to upgrade to Swift 3.1 and also move to the new implementation of the  associated Watson SDKs.  A PG build indicates this may exacerbate the existing startup cost for the first swift3action container tests.

**Comment by csantanapr** [comment]
@paulcastro I want to do the change gradually, I purposely didn't update to swift 3.1 and updated the WatsonSDK and JSON lib to latest. 
I think that should be done in a separate PR

Also @rabbah I think we need to deprecate the use of a single swift file for swift actions.
I think we should restrict to only accepting swift zip precompiled actions. 
This will avoid users to hit a very large cold start by mistake, and also reduce the docker image size were we can use `ibmcon/swift-ubuntu-runtime` and cut the size of the image.

**Comment by rabbah** [comment]
we can debate that separately: if we add the _init_ actions, then users can get that for free.

**Comment by csantanapr** [comment]
That was my thinking also, but from the perspective of the runtime it only takes zip
Even it means the unit actions are helpers to get the zip created/refreshed

**Comment by csantanapr** [comment]
@paulcastro 
When you have a chance can you try this PR.
The image is not being built correctly
This step:
```
 python /swift3Action/buildandrecord.py
Building action
action built. Decoding compile and link commands
Cannot generate build script: compile or link command not found
```

**Comment by paulcastro** [comment]
I'm testing it now - good catch.  I noticed that the home directory gets set to the swift runtime instead of "/"

**Comment by csantanapr** [comment]
fetch again I made an update to make it `/`

the ibmcon/swift-ubuntu image sets the workdir to /root

**Comment by csantanapr** [comment]
@paulcastro https://github.com/openwhisk/openwhisk/pull/2085/commits/bb9d21037e7c7d21f67f83ff3c5b0e265b3f9a10

**Comment by paulcastro** [comment]
`swiftc` is in a different location so lines 26 and 27 of buildandrecord.py should be:

```
COMPILE_PREFIX = "/root/swift-3.0.2-RELEASE-ubuntu14.04/usr/bin/swiftc -module-name Action "
LINKER_PREFIX = "'/root/swift-3.0.2-RELEASE-ubuntu14.04/usr/bin/swiftc' -Xlinker '-rpath=$ORIGIN' '-L/swift3Action/spm-build/.build/release' -o '/swift3Action/spm-build/.build/release/Action'"
```

I have an example in my 3.1 [branch](https://github.com/paulcastro/openwhisk/blob/issue_2079b-x/core/swift3Action/buildandrecord.py)

**Comment by csantanapr** [comment]
Thanks @paulcastro that did the trick https://github.com/openwhisk/openwhisk/pull/2085/commits/33ef2959b69bea6d8873924eb99b89578b38d5a5

**Comment by rabbah** [comment]
I will review - but going to declare this is depend on 2070 to make sure we're not again regressing tests I just fixed.

File: core/swift3Action/Dockerfile [hunk 1]
```diff
@@ -1,51 +1,15 @@
...
-    SWIFT_URL=https://swift.org/builds/swift-$SWIFT_VERSION-$(echo "$SWIFT_RELEASE_TYPE" | tr '[:upper:]' '[:lower:]')/$(echo "$SWIFT_PLATFORM" | tr -d .)/swift-$SWIFT_VERSION-$SWIFT_RELEASE_TYPE/$SWIFT_ARCHIVE_NAME.tar.gz && \
-    echo $SWIFT_URL && \
-    wget --no-verbose $SWIFT_URL && \
-    wget --no-verbose $SWIFT_URL.sig && \
-    gpg --verify $SWIFT_ARCHIVE_NAME.tar.gz.sig && \
-    tar -xzf $SWIFT_ARCHIVE_NAME.tar.gz --directory / --strip-components=1 && \
-    rm -rf $SWIFT_ARCHIVE_NAME* /tmp/* /var/tmp/*
+# Upgrade and install basic Python dependencies
+RUN apt-get -y update \
+ && apt-get -y install --fix-missing python2.7 python-gevent python-flask zip
```
Comment (rabbah) [inline core/swift3Action/Dockerfile]: > python2.7 is already the newest version.
> python2.7 set to manually installed.

can skip python2.7 it's already in the image.

File: core/swift3Action/buildandrecord.py [hunk 2]
```diff
@@ -23,10 +23,8 @@
 from subprocess import check_output
 
 # Settings
-COMPILE_PREFIX = "/usr/bin/swiftc -module-name Action "
-LINKER_PREFIX = "/usr/bin/swiftc -Xlinker '-rpath=$ORIGIN' " \
-                "'-L/swift3Action/spm-build/.build/release' " \
-                "-o '/swift3Action/spm-build/.build/release/Action'"
+COMPILE_PREFIX = os.environ["SWIFT_LOCATION"]+"/usr/bin/swiftc -module-name Action "
+LINKER_PREFIX = "'"+os.environ["SWIFT_LOCATION"]+"/usr/bin/swiftc' -Xlinker '-rpath=$ORIGIN' '-L/swift3Action/spm-build/.build/release' -o '/swift3Action/spm-build/.build/release/Action'"
```
Comment (markusthoemmes) [inline core/swift3Action/buildandrecord.py]: Want to use string interpolation to make this easier to grasp?

```
LINKER_PREFIX = "'%s/usr/bin/swiftc' -Xlinker '-rpath=$ORIGIN' '-L/swift3Action/spm-build/.build/release' -o '/swift3Action/spm-build/.build/release/Action'" % os.environ["SWIFT_LOCATION"]

**Comment by csantanapr** [comment]
@paulcastro going to take care of this when he updates to swift 3.1

**Comment by rabbah** [comment]
FYI I rebased this and fixed a missing slash:

```
- && apt-get -y install --fix-missing python-gevent python-flask
+ && apt-get -y install --fix-missing python-gevent python-flask \
```

one test fails
screen shot 2017-04-03 at 2 58 30 pm

it's because of this error:
```
"/swift3Action/spm-build/.build/release/Action: error while loading shared libraries: libswiftCore.so: cannot open shared object file: No such file or directory" was not empty (SwiftActionContainerTests.scala:235)
```

@paulcastro FYI if it's helpful.

**Comment by csantanapr** [comment]
Yeah I rebase and added a slash on branch @paulcastro https://github.com/csantanapr/openwhisk/commit/91555aaaa3c3b416af56c2415d9fc984c2890ebb

**Comment by paulcastro** [comment]
Updated helloSwift.zip file in `tests/dat/actions` should fix the test.

**Comment by markusthoemmes** [comment]
@csantanapr can we revive this seeing the latest build breakages due to swift downloads directly from swift.org.

**Comment by csantanapr** [comment]
@markusthoemmes this PR is not good as I'm afraid it will break pre-compiled zip actions. 
I think you already took care by pinning the publish image

**Pull request closed without merging**